### PR TITLE
[GNA] Enable CNN2D tests for GNA Lib 2.1.0.1048

### PR DESCRIPTION
**Pull request by kbruniec**

### CODE REVIEW

File: inference-engine/src/gna_plugin/gna_device.cpp [hunk 1]
```diff
@@ -102,23 +115,10 @@ void GNADeviceHelper::enforceLegacyCnns(Gna2Model& gnaModel) {
...
-    }
-    return buffer;
-#endif
-}
-
 uint32_t GNADeviceHelper::createModel(Gna2Model& gnaModel) const {
     uint32_t modelId;
     if (isUpTo20GnaDevice()) {
-        enforceLegacyCnns(gnaModel);
+        // enforceLegacyCnns(gnaModel);
```
Comment (kbruniec) [inline inference-engine/src/gna_plugin/gna_device.cpp:121]: Refine

File: inference-engine/src/gna_plugin/backend/am_intel_dnn.cpp [hunk 2]
```diff
@@ -1489,6 +1532,30 @@ void GNAPluginNS::backend::AMIntelDNN::InitGNAStruct(intel_nnet_type_t *ptr_nnet
                     pConvolutionalLayer->pFilters = component[i].op.conv1D.ptr_filters;
                 }
                 AdvanceCnnOperationIfAllApplied(component, i, pLayer);
+#endif
+                break;
+            case kDnnConvolutional2dOp:
+#if  GNA_LIB_VER == 2
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/src/gna_plugin/backend/am_intel_dnn.cpp:1538]: Its 2D convolution so is available not only on GNA 3.0, but also on future GNA 3.5.
This new name **kDnnConvolutional2dOp** corespondes to the existing one: **kDnnConvolutional1dOp**.

If you meant GNA_LIB_VER then the its definition will be also == 2. GNA library for GNA 3.0/3.5 don't need to be distinguished based on conditional compilation.

File: inference-engine/src/gna_plugin/backend/am_intel_dnn.cpp [hunk 3]
```diff
@@ -979,7 +1016,12 @@ void GNAPluginNS::backend::AMIntelDNN::WriteDnnText(const char *filename, intel_
                     }
                     out_file << "\n";
                 }
-                    break;
+                break;
+                case kDnnConvolutional2dOp: {
+                    fprintf(stderr, "Unsupported kDnnConvolutional2dOp type in WriteDnnText!\n");
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/src/gna_plugin/backend/am_intel_dnn.cpp:1021]: OK

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 4]
```diff
@@ -511,6 +511,234 @@ void GNAGraphCompiler::ConvolutionPrimitive(InferenceEngine::CNNLayerPtr layer)
     }
 }
 
+#if GNA_LIB_VER == 2
+template <typename T, typename F>
+T safeCastTo(F from) {
+    return static_cast<T>(from);
+}
+
+void GNAGraphCompiler::ConvolutionNewPrimitive(InferenceEngine::CNNLayerPtr layer) {
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/src/gna_plugin/gna_graph_compiler.cpp:520]: Already done locally, will push today

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 5]
```diff
@@ -511,6 +511,234 @@ void GNAGraphCompiler::ConvolutionPrimitive(InferenceEngine::CNNLayerPtr layer)
...
+    if (out_channels != convolution._out_depth) {
+        THROW_GNA_LAYER_EXCEPTION(layer) << "Output channels do not equal output depth. "
+            << out_channels << " vs " << convolution._out_depth;
+    }
+
+    if (dnn->new_num_conv_columns) {
+        dnn->new_num_conv_columns = 0;
+    }
+
+    // TODO: BRUNO replace
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/src/gna_plugin/gna_graph_compiler.cpp:570]: Already cleaned now locally, will push today.

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 6]
```diff
@@ -257,22 +258,21 @@ void GNAGraphCompiler::ConvolutionPrimitive(InferenceEngine::CNNLayerPtr layer)
...
+    if (((in_channels > 1) && (in_height > 1) && (in_width > 1)) || // 3D input
+        (convolution._kernel_x != 1 && convolution._kernel_y != 1 && convolution._kernel_y != in_channels) || // 2D kernel
+        (inputLayout != Layout::NHWC && in_height != 1)) {
         // TensorFlow default layout is NHWC
         // OpenVino Default layout is   NCHW
         // GNA Convolution input is     NHCW
         // When layer layout is in NHWC it means that is was created by PassManager
-        THROW_GNA_LAYER_EXCEPTION(layer) << "in_height != 1 This case requires additional Permute and it is not implemented yet";
+#if GNA_LIB_VER == 2
+        return ConvolutionNewPrimitive(layer);
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/src/gna_plugin/gna_graph_compiler.cpp:273]: already done locally, will push today

File: inference-engine/src/gna_plugin/gna_device.cpp (hunk 1 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/gna_device.cpp]: done

File: inference-engine/src/gna_plugin/backend/am_intel_dnn.hpp [hunk 7]
```diff
@@ -133,6 +133,34 @@ class AMIntelDNN {
                                             true);
     }
 
+#if GNA_LIB_VER == 2
```
Comment (bsochack) [inline inference-engine/src/gna_plugin/backend/am_intel_dnn.hpp:136]: There is no better place, so I will leave this comment here.
gna_float_runtime.cpp file contains a reference implementation of all transformations. I guess it is used by FP32 mode.
Did you test FP32 mode?

Comment (kbruniec) [reply inference-engine/src/gna_plugin/backend/am_intel_dnn.hpp:136]: FP32 mode is probably not tested.
I would create separate ticket for implementation and testing of "reference CNN2D FP32 transformation" or we risk newer-ending story.
What do you think?

Comment (kbruniec) [reply inference-engine/src/gna_plugin/backend/am_intel_dnn.hpp:136]: @bsochack Could you comment/resolve this issue and possibly approve this PR?

**Review by bsochack — APPROVED** [review]

**Pull request merged**